Suggest for loop instead of while-let when looping over iterators

fhartwig opened the pull request:
Fixes #387
I think this is ready for review. I even remembered to update the readme for once :smile:

fhartwig commented:
Argh. I just realised that I misread https://github.com/rust-lang/rust/issues/8372 . This issue means that there are cases where `while let` is actually what you want (I thought that `for` had been fixed`). I guess I will have to also check that the iterator isn't used inside the loop body to avoid false positives.

Manishearth commented on tests/compile-fail/while_loop.rs:
```diff
@@ -53,6 +53,23 @@ fn main() {
     }
+
+
+    while let Option::Some(x) = (1..20).next() { //~ERROR this loop could be written as a `for` loop
```
Since these examples depend on a temporary, they are all actually infinite loops :smile: 

I'd prefer if we used vector iteration here or stored the ranges in an external mutable variable.

fhartwig replied:
I'm not sure what I was thinking there :laughing: Thanks, I'll fix that.

Manishearth commented on src/loops.rs:
```diff
@@ -228,6 +231,17 @@ impl LateLintPass for LoopsPass {
+            if let (&PatEnum(ref path, _), &ExprMethodCall(method_name, _, _)) = (pat, &expr.node) {
+                if method_name.node.as_str() == "next" &&
+                        match_trait_method(cx, expr, &["core", "iter", "Iterator"]) &&
+                        path.segments.last().unwrap().identifier.name.as_str() == "Some" {
```
Use an if let instead of unwrapping here

fhartwig replied:
I can do that if you prefer, but let me defend `unwrap` here for a second: As far as I can tell, `path.segments` should always be non-empty, so `last` should always return a `Some`. Now, I might be wrong about that, but if I am I'd rather have the program crash instead of silently carry on in some state that I didn't anticipate. (Ok, I should still use `expect` so clippy at least crashes with a sensible error message.)

Manishearth replied:
Expect is fine.

Note that when a plugin crashes, it looks like an ICE, so really we should be using `span_bug` or something here with an if let.

(Also, the CI won't accept a clippy PR which emits clippy warnings, run `bash util/dogfood.sh` to see)

llogiq commented on src/loops.rs:
```diff
@@ -228,6 +231,17 @@ impl LateLintPass for LoopsPass {
+                        match_trait_method(cx, expr, &["core", "iter", "Iterator"]) &&
+                        path.segments.last().unwrap().identifier.name.as_str() == "Some" {
+                    span_lint(cx, WHILE_LET_ON_ITERATOR, expr.span,
+                              "this loop could be written as a `for` loop");
```
An obvious extension would be to extract the relevant parts of the code and rewrite the loop as a `for` loop using snippet.

fhartwig replied:
Good idea! I'll do that.

llogiq replied:
Great work so far. Tell us if you need anything.

fhartwig commented:
I've made fixes for your comments, fixed the false positives where the iterator is used inside the loop body, and rebased the branch onto the current master. Ready for more feedback :smile:

llogiq commented on README.md:
```diff
@@ -6,7 +6,8 @@ A collection of lints to catch common mistakes and improve your Rust code.
 ##Lints
-There are 68 lints included in this crate:
+There are 69 lints included in this crate:
+There are 65 lints included in this crate:
```
This looks wrong.

fhartwig replied:
Looks like I messed up when resolving a conflict for the rebase. Fixed.

llogiq replied:
Happens to the best of us :-)

llogiq commented on src/loops.rs:
```diff
@@ -300,6 +326,32 @@ impl<'v, 't> Visitor<'v> for VarVisitor<'v, 't> {
+    }
+}
+
+struct VarUsedVisitor<'v, 't: 'v> {
```
This is quite a useful thing to have – I imagine we could use it in the shadows lint with a bit of tweaking.

llogiq commented on tests/compile-fail/while_loop.rs:
```diff
@@ -53,6 +53,33 @@ fn main() {
+    let mut iter = 1..20;
+    while let Some(x) = iter.next() { //~ERROR this loop could be written as a `for` loop
+        println!("{}", x);
+    }
```
I guess it's a contrived example, but what if the loop breaks and the `iter` is reused? I think `for` loops usually consume the iterator, am I right?

We should at least have a test case.

Manishearth replied:
iirc they borrow

fhartwig replied:
No, for loops consume the iterator (probably because `IntoIterator::into_iter` takes `self` by value). I actually considered that, but couldn't come up with convincing use-cases where you would want to use the iterator after the loop. On the other hand, it shouldn't be very hard to fix.

Manishearth replied:
Huh, you're right.

llogiq replied:
Whew! You had me worried I was mis-remembering how iterators worked – and I already had three cups of coffee, so my brain should sort of work currently... :smile:

Anyway, yeah, it shouldn't be too hard to detect and omit the lint in those situations.

fhartwig commented:
Regarding the discussion about iterators being used after the while-let loop:

> On the other hand, it shouldn't be very hard to fix.

Ok, I have to eat my words here, I'm a bit stumped. I want to have a function that checks whether the an iterator variable is used again after being used in a while-let loop. I have written a visitor that, given a block, the iterator variable and the id of the while-let loop, checks whether the iterator variable appears after the beginning of the while let loop (i.e. in the loop body or after the loop). But now I need a block to run it on. I thought I could just use `get_parent_node` with the node id of the variable declaration to get the relevant block, but the parent node isn't necessarily a `NodeBlock`, it might be a `NodeItem` or a `NodeImpltem` or whatever `Node` variant can contain a `Block`. I'm guessing there must be some straightforward way to get the block (or a list of statements or something along those lines) that a given binding is in scope, but I can't seem to find it.

fhartwig commented:
Ok, I just found `get_enclosing_scope` in rustc, I guess that's something that could be useful.

Manishearth commented:
Note: for tracking uses, try `ExprUseVisitor`

fhartwig commented:
I've fixed the false positives when an iterator is used after the loop. I've not used `ExprUseVisitor` since that ended up being more complex than what I do now. On the other hand, `ExprUseVisitor` would catch certain false negatives, e.g. the following:
```rust
let mut iter = 0u32..20;
while let Some(x) = iter().next {}
iter = 1..5;
println!("iter: {:?}" iter);
```
This should strictly speaking warn since it can be written as a for loop, but it currently doesn't since `iter` is used after the loop. I think this is too much of an edge case to make the code more complicated for it, but I'm not sure.

Regarding the `dogfood.sh` failure: I think this is a false positive in the `needless_lifetime` lint, because when I remove the lifetime, rustc complains about missing lifetimes. But it's late now and I'll look into it tomorrow.

Manishearth commented:
```
 pub fn get_enclosing_block(cx: &LateContext, node: NodeId) -> Option<&Block> {
```

should work, really.

fhartwig commented:
@Manishearth That's what I thought, but it doesn't (Probably because of the lifetimes on `LateContext`? I don't really know the elision rules off the top of my head). So as far as I can tell, this is a false positive. I've just filed #417

Manishearth commented:
Oh, LateContext has a tcx lifetime. Right.

Manishearth commented:
Since you had done most of the review, r? @llogiq 

(If you're busy let me know and I'll do it tomorrow)

llogiq commented:
Looks good to me; just slap an `#[allow(needless_lifetimes)]` on the offending function until we've fixed #417, and I'll be happy to merge.

fhartwig commented:
Done :smile:

llogiq commented:
Thanks a bunch! :smiley:
